rundeckapp/src/main/groovy/rundeck/services/audit/AuditEventsService.groovy

```diff
@@ -0,0 +1,534 @@
+     *         this allows developers to set-up a "base" builder, and then modify it and publish many times,
+     *         without compromising the data on past events.
+     *     </p>
+     *
+     *
+     */
+    final class AuditEventBuilder {
+
+        private String username = null
+        private Collection<String> userRoles = []
+        private String action = null
+        private String resourceType = null
+        private String resourceName = null
+        private GrailsWebRequest httpRequest = null;
+
+        private AuditEventBuilder() {
+        }
+
+        /**
+         * Triggers this event publishing.
+         * A new event will be generated and published with an immutable copy of the data in this builder.
+         * Multiple calls to this method will cause multiple events to be published.
+         * <p>
+         *     If the username is not set, the user and roles will be set
+         *     from the current security context.
+         */
+        void publish() {
+            dispatchEvent(this)
+        }
+
+        /**
+         * Sets the username associated with this event.
+         * If not set, the current context user will be used.
+         * @param username
+         * @return
+         */
+        AuditEventBuilder setUsername(String username) {
+            this.username = username
+            return this
+        }
+
+        /**
+         * Sets the roles of the user.
+         * @param roles
+         * @return
+         */
+        AuditEventBuilder setUserRoles(List<String> roles) {
+            this.userRoles = roles
+            return this
+        }
+
+        /**
+         * Sets the event action.
+         * See {@link ActionTypes}
+         * @param eventType
+         * @return
+         */
+        AuditEventBuilder setActionType(String actionType) {
+            this.action = actionType
+            return this
+        }
+
+        /**
+         * Sets the resource type associated with this event.
+         * @param projectName
+         * @return
+         */
+        AuditEventBuilder setResourceType(String resourceType) {
+            this.resourceType = resourceType
+            return this
+        }
+
+        /**
+         * Sets the resource name or ID associated with this event.
+         * @param projectName
+         * @return
+         */
+        AuditEventBuilder setResourceName(String name) {
+            this.resourceName = name
+            return this
+        }
+
+
+        /**
+         * Sets the request associated with this event.
+         * @param httpRequest HTTP Request associated with this event.
+         * @return
+         */
+        AuditEventBuilder setHttpRequest(GrailsWebRequest httpRequest) {
+            this.httpRequest = httpRequest
+            return this
+        }
+
+
+        /**
+         * Builds a new immutable event object
+         * @return
+         */
+        protected AuditEvent build() {
```

Comment:
this is a lot of boiler-plate builder code, perhaps use Lombok `@Builder` in a core Java class?

Reply:
I can refactor the intrface implementation to a separate POJO, but i still need some of that boilerplate to ensure deep data copy between objects and to get correctly the request context.